File: src/CAMspec/clik_CAMspec.c
#include "clik.h"
#include "clik_helper.h"
#include <errno.h>
#include <string.h>

void free_CAMspec(void **none) {
  camspec_extra_free_();
}

double CAMspec_lkl(void* none, double* pars, error **err) {
  double lkl;
  
  camspec_extra_lkl_(&lkl,pars);
  return lkl;
}

void camspec_extra_init_(int*, int*,int*,int*,int*,int*, double*,double*,int*,double*,double*,double*,int*,int*,int*,int*,double*,double*);
//CAMSPEC_EXTRA_INIT(iNspec, inX,ilminX,ilmaxX,inp,inpt, ic_inv,iX,iX,ilmax_sz,isz_143_temp,iksz_temp,itszxcib_temp,ibeam_Nspec,inum_modes_per_beam,ibeam_lmax,icov_dim,ibeam_cov_inv,ibeam_modes)

cmblkl* clik_CAMspec_init(cldf *df, int nell, int* ell, int* has_cl, double unit,double* wl, double *bins, int nbins, error **err) {
  hsize_t ndum;
  int bok;
  cmblkl *cing;
  herr_t hstat;
  char likefile[2048],szfile[2048];
  int xcase;
  double xdim;
  char *xnames_def[] = {"A_ps_100","A_ps_143", "A_ps_217", "A_cib_143", "A_cib_217", "A_sz",  
                      "r_ps", "r_cib","n_Dl_cib","cal_100", "cal_143", "cal_217","xi_sz_cib","A_ksz"};
  
  char *xnames_1[] = {"A_ps_100","A_ps_143", "A_ps_217", "A_cib_143", "A_cib_217", "A_sz",  
                      "r_ps", "r_cib","cal0", "cal1", "cal2"};
  char *xnames_0[] = {"A_ps_143", "A_cib_143", "A_sz",  
                      "cal1", "cal2"};

  char *xnames_tot[300];

  int Nspec, nX, lmax_sz;
  int *lmaxX, *lminX, *np, *npt;
  double *X,*c_inv,*tsz,*ksz,*tszXcib;
  int X_sz,c_inv_sz,sz_temp_sz;
  int beam_Nspec,num_modes_per_beam,beam_lmax,cov_dim;
  double *beam_cov_inv,*beam_modes;
  int beam_cov_inv_sz,beam_modes_sz;
  int i,j,cnt;
  //int frq[] = {100,143,217};
  
  camspec_extra_only_one_(&bok);
  testErrorRet(bok!=0,-100,"CAMspec already initialized",*err,__LINE__,NULL);
  
  Nspec = cldf_readint(df,"Nspec",err);
  forwardError(*err,__LINE__,NULL);
  
  lminX = cldf_readintarray(df,"lminX",&Nspec,err);
  forwardError(*err,__LINE__,NULL);
  lmaxX = cldf_readintarray(df,"lmaxX",&Nspec,err);
  forwardError(*err,__LINE__,NULL);

  np = cldf_readintarray(df,"np",&Nspec,err);
  forwardError(*err,__LINE__,NULL);
  npt = cldf_readintarray(df,"npt",&Nspec,err);
  forwardError(*err,__LINE__,NULL);
 
  nX = cldf_readint(df,"nX",err);
  forwardError(*err,__LINE__,NULL);

  lmax_sz = cldf_readint(df,"lmax_sz",err);
  forwardError(*err,__LINE__,NULL);

  X_sz = -1;
  X = cldf_readfloatarray(df,"X",&X_sz, err);
  forwardError(*err,__LINE__,NULL);
  
  c_inv_sz = -1;
  c_inv = cldf_readfloatarray(df,"c_inv",&c_inv_sz, err);
  forwardError(*err,__LINE__,NULL);

  sz_temp_sz = -1;
  tsz = cldf_readfloatarray(df,"tsz",&sz_temp_sz, err);
  forwardError(*err,__LINE__,NULL);
  sz_temp_sz = -1;
  ksz = cldf_readfloatarray(df,"ksz",&sz_temp_sz, err);
  forwardError(*err,__LINE__,NULL);
  sz_temp_sz = -1;
  tszXcib = cldf_readfloatarray(df,"tszXcib",&sz_temp_sz, err);
  forwardError(*err,__LINE__,NULL);


  beam_Nspec = cldf_readint(df,"beam_Nspec",err);
  forwardError(*err,__LINE__,NULL);

  if (beam_Nspec==0) {
    num_modes_per_beam = 0;
    beam_lmax = 0;
    cov_dim   = 0;
    beam_cov_inv = malloc_err(sizeof(double)*1,err);
    forwardError(*err,__LINE__,NULL);
    beam_modes = malloc_err(sizeof(double)*1,err);
    forwardError(*err,__LINE__,NULL);
  } else {
    beam_lmax = cldf_readint(df,"beam_lmax",err);
    forwardError(*err,__LINE__,NULL);
    num_modes_per_beam = cldf_readint(df,"num_modes_per_beam",err);
    forwardError(*err,__LINE__,NULL);
    cov_dim = cldf_readint(df,"cov_dim",err);
    forwardError(*err,__LINE__,NULL);
  
    beam_cov_inv_sz = -1;
    beam_cov_inv = cldf_readfloatarray(df,"beam_cov_inv",&beam_cov_inv_sz, err);
    forwardError(*err,__LINE__,NULL);
    beam_modes_sz = -1;
    beam_modes = cldf_readfloatarray(df,"beam_modes",&beam_modes_sz, err);
    forwardError(*err,__LINE__,NULL);
    
  }
  camspec_extra_init_(&Nspec, &nX,lminX,lmaxX,np,npt,c_inv,X,&lmax_sz, tsz,ksz,tszXcib,&beam_Nspec,&num_modes_per_beam,&beam_lmax,&cov_dim,beam_cov_inv,beam_modes);    
  
  
  //camspec_extra_getcase_(&xcase);
  xdim = 14 + beam_Nspec*num_modes_per_beam;
  
  for(i=0;i<14;i++) {
    xnames_tot[i] = xnames_def[i];
  }

  cnt = 14;
  
  for(i=0;i<beam_Nspec;i++) {
    for(j=0;j<num_modes_per_beam;j++) {
      xnames_tot[cnt] = malloc_err(sizeof(char)*50,err);
      forwardError(*err,__LINE__,NULL);
      sprintf(xnames_tot[cnt],"Bm_%d_%d",i+1,j+1);
      cnt++;
    }
  }

  cing = init_cmblkl(NULL, &CAMspec_lkl, 
                     &free_CAMspec,
                     nell,ell,
                     has_cl,ell[nell-1],unit,wl,0,bins,nbins,xdim,err);
  forwardError(*err,__LINE__,NULL);
  /*if (xcase==0) {
    cmblkl_set_names(cing, xnames_0,err);
  forwardError(*err,__LINE__,NULL);  
  } else {*/
  cmblkl_set_names(cing, xnames_tot,err);
  forwardError(*err,__LINE__,NULL);
  //}
  
  free(X);
  free(c_inv);
  free(ksz);
  free(tsz);
  free(tszXcib);
  free(beam_modes);
  free(beam_cov_inv);
  for(i=14;i<xdim;i++) {
    free(xnames_tot[i]);
  }

  return cing;
}

double* camspec_get_fg(void* camclik,double *par,int lmax,error **err) {
  double *res;

  res = malloc_err(sizeof(double)*(lmax+1)*4,err);
  forwardError(*err,__LINE__,NULL);

  camspec_extra_fg_(res,par,&lmax);

  return res;
}

File: src/CAMspec/clik_CAMspec.f90
MODULE CAMSPEC_EXTRA

    IMPLICIT NONE

    INTEGER:: BOK = 0
    real(8), dimension(:), allocatable :: cltt
    INTEGER::lmin,lmax
    INTEGER::xcase,npar
END MODULE CAMSPEC_EXTRA



SUBROUTINE CAMSPEC_EXTRA_ONLY_ONE(MOK)
    USE CAMSPEC_EXTRA
    INTEGER,INTENT(OUT)::MOK
    MOK = BOK
    BOK = 1
END SUBROUTINE  CAMSPEC_EXTRA_ONLY_ONE

SUBROUTINE CAMSPEC_EXTRA_FREE()
    USE CAMSPEC_EXTRA
    DEALLOCATE(cltt)
    BOK =0
END SUBROUTINE  CAMSPEC_EXTRA_FREE

SUBROUTINE CAMSPEC_EXTRA_INIT(iNspec, inX,ilminX,ilmaxX,inp,inpt,ilmax_sz, sz_100,sz_143,mc_inv,mX)
    USE CAMSPEC_EXTRA
    USE temp_like
    implicit none
    integer, intent(in):: iNspec, iNx,ilmax_sz
    integer,dimension(1:iNspec),intent(in)::ilmaxX, ilminX,inp,inpt
    real(8),dimension(0:ilmax_sz),intent(in)::sz_100,sz_143
    real(8),dimension(1:iNx),intent(in)::mX
    real(8),dimension(1:iNx,1:iNx),intent(in)::mc_inv

    integer::i
    
    call like_init_frommem(iNspec, inX,ilminX,ilmaxX,inp,inpt,ilmax_sz, sz_100,sz_143,mc_inv,mX)
    
    lmax = ilmaxX(1)
    DO i=2,iNspec
        IF (lmax<ilmaxX(i)) lmax = ilmaxX(i)
    ENDDO
    lmin = ilminX(1)
    DO i=2,iNspec
        IF (lmin>ilminX(i)) lmin = ilminX(i)
    ENDDO
    
    xcase = 1
    !IF (mlmax2.ne.0) xcase = 1

    ALLOCATE(cltt(0:lmax+1))
    npar = lmax+1-lmin+11
    
END SUBROUTINE CAMSPEC_EXTRA_INIT

SUBROUTINE CAMSPEC_EXTRA_GETCASE(xc)
    USE CAMSPEC_EXTRA
    INTEGER::xc

    xc = xcase
END SUBROUTINE CAMSPEC_EXTRA_GETCASE

SUBROUTINE CAMSPEC_EXTRA_LKL(LKL,CL)
    USE CAMSPEC_EXTRA
    use temp_like
    implicit none

    REAL(8),INTENT(OUT)::LKL
    REAL(8),DIMENSION(0:npar-1)::CL
    real(8)::zlike, A_ps_100, A_ps_143, A_ps_217, A_cib_143, A_cib_217, A_sz, r_ps, r_cib, &
         cal0, cal1, cal2
    INTEGER::l
    real(8)::tlkl

    cltt(:lmin-1) = 0
    DO l=lmin,lmax
        ! camspec expects cl/2pi !!! argl !
        cltt(l)=CL(l-lmin)/2./3.14159265358979323846264338328
    ENDDO

    A_ps_100 = CL(lmax+1-lmin + 0)
    A_ps_143 = CL(lmax+1-lmin + 1)
    A_ps_217 = CL(lmax+1-lmin + 2)
    A_cib_143 = CL(lmax+1-lmin + 3)
    A_cib_217 = CL(lmax+1-lmin + 4)
    A_sz      = CL(lmax+1-lmin + 5)
    r_ps      = CL(lmax+1-lmin + 6)
    r_cib     = CL(lmax+1-lmin + 7) 
    cal0      = CL(lmax+1-lmin + 8) 
    cal1      = CL(lmax+1-lmin + 9) 
    cal2      = CL(lmax+1-lmin + 10)     

    call calc_like(tlkl,  cltt, A_ps_100, A_ps_143, A_ps_217, A_cib_143, A_cib_217, A_sz, r_ps, r_cib, cal0,cal1, cal2)
    ! lkl is -2loglike clik returns loglik
    !print *,tlkl
    lkl = -tlkl/2.

END SUBROUTINE CAMSPEC_EXTRA_LKL

